mldsa/rounding.h

```diff
@@ -76,6 +76,11 @@ __contract__(
 );
 
 #define use_hint MLD_NAMESPACE(use_hint)
-int32_t use_hint(int32_t a, unsigned int hint);
+int32_t use_hint(int32_t a, unsigned int hint)
+__contract__(
+  requires(hint >= 0 && hint <= 1)
+  requires(a >= 0 && a < MLDSA_Q)
+  ensures(return_value >= 0 && return_value < (MLDSA_Q-1)/(2*MLDSA_GAMMA2))
```

Comment:
Based on Algorithm 40 in FIPS204, I believe the `return_value` should fall within the range `0 <= return_value <= (q-1)/(2*MLDSA_GAMMA2)`. To properly specify this constraint in the code, we would need to change the post-condition check to `return_value <=(MLDSA_Q-1)/(2*MLDSA_GAMMA2)`.
```suggestion
  ensures(return_value >= 0 && return_value <= (MLDSA_Q-1)/(2*MLDSA_GAMMA2))
```

Reply:
I wrote and exhaustive test case for decompose(), and it only ever returns a1 in 0 .. 15 (inclusive) when MLDSA_MODE=3 or 0 .. 43 when MLDSA_MODE=2, so it looks like the reference implementation has a slightly stronger postcondition that the specification.

Reply:
The code is ugly (it's doing shifts and bitwise operations on signed integers which are, strictly speaking, implementation-defined), but if we assume the "obvious" 2's complement representation, then you can see that the final assignment to a1 in decompose() can only ever yield 0..15 or 0..43 for each case.

Reply:
I just added some intermediate assertions in decompose() that justify the final range of *a1
That, in turn, explains the post-condition of use_hint()

Does that make sense?